### PR TITLE
[SPARK-36500][CORE] Fix temp_shuffle file leaking when a task is interrupted

### PR DESCRIPTION
**Pull request by jiangxb1987**
### What changes were proposed in this pull request?

When a task thread is interrupted, the underlying output stream referred by `DiskBlockObjectWriter.mcs` may have been closed, then we get IOException when flushing the buffered data. This breaks the assumption that `revertPartialWritesAndClose()` should not throw exceptions.

To fix the issue, we can catch the IOException in `ManualCloseOutputStream.manualClose()`.

### Why are the changes needed?

Previously the IOException was not captured, thus `revertPartialWritesAndClose()` threw an exception. When this happens, `BypassMergeSortShuffleWriter.stop()` would stop deleting the temp_shuffle files tracked by `partitionWriters`, hens lead to temp_shuffle file leak issues.

### Does this PR introduce _any_ user-facing change?

No, this is an internal bug fix.

### How was this patch tested?

Tested by running a longevity stress test. After the fix, there is no more leaked temp_shuffle files.

### CODE REVIEW

**Comment by jiangxb1987** [comment]
cc @Ngone51 @cloud-fan @xuanyuanking

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thanks, @jiangxb1987 .

File: core/src/main/scala/org/apache/spark/storage/DiskBlockObjectWriter.scala [hunk 1]
```diff
@@ -65,7 +65,16 @@ private[spark] class DiskBlockObjectWriter(
     }
 
     def manualClose(): Unit = {
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/storage/DiskBlockObjectWriter.scala:67]: Ideally, we should handle this in `closeResources` - but it is fine here as well, given this is not `close` but `manualClose` and we control the semantics of it :-)

**Review by mridulm — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
@jiangxb1987 would you mind enabling Github actions in your fork (see https://github.com/apache/spark/pull/33731/checks?check_run_id=3316027997)?

**Review by HyukjinKwon — APPROVED** [review]

**Comment by jiangxb1987** [comment]
@HyukjinKwon I selected "Allow all actions" and clicked the Save button, but the Workflow run detection still fail. As you can see, the base branch is also the same as lastest master.

**Review by Ngone51 — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master, branch-3.2, branch-3.1 and brnach-3.0.

**Comment by jiangxb1987** [comment]
Thank @mridulm @HyukjinKwon @dongjoon-hyun @Ngone51 !

**Pull request closed without merging**